File: net/src/Models/Defense/Defense.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using System.IO;
using System.Linq;
using System.Web;

namespace TestDefense.Models.Defense
{
    public class AttackDescription {
        public string User;
        public string Ip;
        public string Attack;
    }

    public class SessionParameters {
        public string User;
        public string Ip;
        public string Cookie;
        public string File;
        public string Uri;
        public string Parameter;

        public SessionParameters(string user, string ip, string cookie, string file, string uri, string parameter) {
            User = user;
            Ip = ip;
            Cookie = cookie;
            File = file;
            Uri = uri;
            Parameter = parameter;
        }
    }

    public class Defense
    {
        public int OK = 1;
        public int ERROR = 0;
        public int ATTACK = -1;
        public int BAN = 100;
        public int MAX_REQUESTS_MINUTE = 100;
        public bool DEBUG = true;
        public int banSeconds = 60 * 60 * 24;
        public string DB = "c:\\temp\\attackers12.db";
        protected SQLiteConnection DBConnection { get; set; }
        public Delegate LogoutMethod { get; set; }


        // Constructor
        public Defense() {
            initDb();
        }

        // Data getters
        protected String ServerVariable(String value) {
            return HttpContext.Current.Request.ServerVariables[value];
        }

        protected String RequestVariable(String key) {
            return HttpContext.Current.Request.Form[key];
        }

        protected object SessionVariable(String key) {
            return HttpContext.Current.Session[key];
        }

        protected void SessionVariableSet(String key, object value) {
            HttpContext.Current.Session[key] = value;
        }

        protected String GetCookie(String key) {
            var value = HttpContext.Current.Request.Cookies[key];
            if (value == null) return null;
            return value.Value;
        }

        protected void SetCookie(String key, String value)
        {
            var cookie = HttpContext.Current.Request.Cookies[key];
            if (cookie == null)
                HttpContext.Current.Request.Cookies.Add(new HttpCookie(key, value));
            else
                cookie.Value = value;
        }

        // Helpers
        protected bool IsSet(object value) {
            return (value != null) && ((String) value != "");
        }

        protected bool IncludesString(string source, string value) {
            return source.ToLower().Contains(value.ToLower());
        }

        // DB Helper methods
        protected void initDb()
        {
            if (DBConnection != null)
                return;

            if (!File.Exists(DB))
            {
                SQLiteConnection.CreateFile(DB);
                DBConnection = new SQLiteConnection("data source=" + DB);
                DBConnection.Open();
                executeNonQuery("CREATE TABLE attacker (id INTEGER PRIMARY KEY, timestamp TEXT, application TEXT, ip TEXT, user TEXT, cookie TEXT, filename TEXT, attack TEXT, score INTEGER)");
                executeNonQuery("CREATE TABLE denyUserAgent (id INTEGER PRIMARY KEY, useragent TEXT)");
                executeNonQuery("INSERT INTO denyUserAgent (useragent) VALUES ('burpcollaborator'), ('dirbuster'), ('nessus'), ('nikto'), ('nmap'), ('paros'), ('python-urllib'), ('qualysguard'), ('sqlmap'), ('useragent'), ('w3af')");
                executeNonQuery("CREATE TABLE denyUrlString (id INTEGER PRIMARY KEY, string TEXT)");
                executeNonQuery("INSERT INTO denyUrlString (string) VALUES ('acunetix'), ('burpcollab'), ('nessus'), ('nikto'), ('parosproxy'), ('qualys'), ('vega'), ('ZAP')");
                executeNonQuery("CREATE TABLE acceptHttpMethod (id INTEGER PRIMARY KEY, method TEXT)");
                executeNonQuery("INSERT INTO acceptHttpMethod (method) VALUES ('HEAD'), ('GET'), ('POST'), ('OPTIONS')");
                executeNonQuery("CREATE TABLE denyExtension (id INTEGER PRIMARY KEY, extension TEXT)");
                executeNonQuery("INSERT INTO denyExtension (extension) VALUES ('bac'), ('BAC'), ('backup'), ('BACKUP'), ('bak'), ('BAK'), ('conf'), ('cs'), ('csproj'), ('inc'), ('INC'), ('ini'), ('java'), ('log'), ('lst'), ('old'), ('OLD'), ('orig'), ('ORIG'), ('sav'), ('save'), ('temp'), ('tmp'), ('TMP'), ('vb'), ('vbproj')");
            }
            else {
                DBConnection = new SQLiteConnection("data source=" + DB);
                DBConnection.Open();
            }
        }

        public void Clear() {
            executeNonQuery("delete from attacker");
        }

        public IList<AttackDescription> getAttacks()
        {
            var results = executeQuery("SELECT id, timestamp, application, ip, user, attack FROM attacker");
            var result = new List<AttackDescription>();

            while (results.Read()) {
                var attack = new AttackDescription() {
                    User = (String) results["user"],
                    Ip = (String) results["ip"],
                    Attack = (String) results["attack"]
                };
                result.Add(attack);
            }
            
            return result;
        }

        protected int executeNonQuery(String query)
        {
            using (var command = new SQLiteCommand(DBConnection))
            {
                command.CommandText = query;
                command.CommandType = CommandType.Text;
                //try
                //{
                return command.ExecuteNonQuery();
                //}
                //catch
                //{
                //return 0;
                //}
            }
        }

        protected SQLiteDataReader executeQuery(String query)
        {
            using (var command = new SQLiteCommand(DBConnection))
            {
                command.CommandText = query;
                command.CommandType = CommandType.Text;
                return command.ExecuteReader();
            }
        }

        // Catch unhandled exceptions
        void exception_handler(Exception exception)
        {
            var str = "Uncaught exception found. Exception message: " + exception.Message + "\n";
            Console.WriteLine(str);
            attackDetected("Uncaught exception", 100);
        }

        void logoutSession()
        {
            if (LogoutMethod != null)
                LogoutMethod.DynamicInvoke();
        }

        public bool isSessionBanned()
        {
            var parameters = getSessionParameters();
            var extra = "";
            var total = (long)0;

            if (IsSet(parameters.User)) extra += "OR user = :user";
            if (IsSet(parameters.Cookie)) extra += "OR cookie = :cookie";

            using (var command = new SQLiteCommand(DBConnection))
            {
                command.CommandText = "SELECT ifnull(SUM(score), 0) AS total FROM attacker WHERE timestamp > :timestamp AND (ip = :ip " + extra + ")";
                command.Parameters.Add(new SQLiteParameter(":timestamp", banSeconds));
                command.Parameters.Add(new SQLiteParameter(":ip", parameters.Ip));
                if (IsSet(parameters.User)) command.Parameters.Add(new SQLiteParameter(":user", parameters.User));
                if (IsSet(parameters.Cookie)) command.Parameters.Add(new SQLiteParameter(":cookie", parameters.Cookie));
                var result = command.ExecuteReader();

                using (result)
                {
                    result.Read();
                    total = (long)result["total"];
                }
            }

            return total >= BAN;
        }

        // Helper methods
        protected DateTime GetTime() {
            return DateTime.Now;
        }

        // Get the session stuff: IP, user (optional), cookie (optional)
        public SessionParameters getSessionParameters()
        {
            string parameter = "", uri = "", file = "", cookie = "", ip = "127.0.0.1", user = "";
            if (IsSet(SessionVariable("user"))) user = (String) SessionVariable("user");
            if (IsSet(ServerVariable("REMOTE_ADDR"))) ip = ServerVariable("REMOTE_ADDR");
            if (IsSet(GetCookie("cookie"))) cookie = GetCookie("cookie");
            if (IsSet(ServerVariable("SCRIPT_FILENAME"))) file = ServerVariable("SCRIPT_FILENAME");
            if (IsSet(ServerVariable("REQUEST_URI"))) uri = ServerVariable("REQUEST_URI");
            if (IsSet(ServerVariable("ALL_RAW"))) parameter = ServerVariable("ALL_RAW");
            return new SessionParameters(user, ip, cookie, file, uri, parameter);
        }

        public void alertAdmin(AlertInfo alert) {
            if (DEBUG) Console.WriteLine("%1-%2-%3-%4\n", alert.Cookie, alert.Ip, alert.User, alert.Description);
        }

        public void alertAdmin(string alert)
        {
            if (DEBUG) Console.WriteLine("%1\n", alert);
        }

        protected void logAttack(String attack, int score)
        {
            using (var command = new SQLiteCommand(DBConnection))
            {
                var parameters = getSessionParameters();
                command.CommandText = "INSERT INTO attacker (timestamp, application, ip, user, cookie, filename, attack, score) VALUES (:timestamp, :application, :ip, :user, :cookie, :filename, :attack, :score)";
                command.Parameters.Add(new SQLiteParameter(":timestamp", GetTime()));
                command.Parameters.Add(new SQLiteParameter(":application", "test-defense.php"));
                command.Parameters.Add(new SQLiteParameter(":ip", parameters.Ip));
                command.Parameters.Add(new SQLiteParameter(":user", parameters.User));
                command.Parameters.Add(new SQLiteParameter(":cookie", parameters.Cookie));
                command.Parameters.Add(new SQLiteParameter(":filename", ServerVariable("SCRIPT_FILENAME")));
                command.Parameters.Add(new SQLiteParameter(":uri", ServerVariable("REQUEST_URI")));
                command.Parameters.Add(new SQLiteParameter(":parameter", HttpContext.Current.Request["ALL_RAW"]));
                command.Parameters.Add(new SQLiteParameter(":attack", attack));
                command.Parameters.Add(new SQLiteParameter(":score", score));
                //try
                //{
                command.ExecuteNonQuery();
                //}
                //catch { }
                //}
            }
        }

        public void attackDetected(String message, int score) {
            logAttack(message, score);
            if (isSessionBanned())
            {
                var description = "The last attack from the user was: " + message + ". ";
                if (score >= BAN)
				    description += "The user was automatically mark as an attacker";
			    else
				    description += "The user was mark as an attacker because of a series of events";

                var parameters = getSessionParameters();
                alertAdmin(new AlertInfo(parameters.Ip, parameters.User, parameters.Cookie, parameters.File, parameters.Uri, parameters.Parameter, description));
            }
        }

        // Checks
        public int checkUserAgent()
        {
            var attack = "Vulnerability scanner in user-agent";
            var score = 100;

            if (IsSet(ServerVariable("HTTP_USER_AGENT")))
                using (var results = executeQuery("SELECT useragent FROM denyUserAgent")) { 
                    while (results.Read())
                        if (IncludesString(ServerVariable("HTTP_USER_AGENT"), (String) results["useragent"]))
                        {
                            attackDetected(attack, score);
                            return ATTACK;
                        }
                }
            else
                return ERROR;

            if (IsSet(SessionVariable("user_agent")) && IsSet(ServerVariable("HTTP_USER_AGENT")))
                if ((String) SessionVariable("user_agent") != ServerVariable("HTTP_USER_AGENT")) {
                    attack = "User-agent changed during user session";
                    attackDetected(attack, score);
                    return ATTACK;
                }

            return OK;
        }

        public int checkConcurrentSession()
        {
            var attack = "The IP address of the user changed for the cookie";
            var score = 25;

            if (IsSet(SessionVariable("REMOTE_ADDR")) && IsSet(ServerVariable("REMOTE_ADDR")))
            {
                if (String.Equals(SessionVariable("REMOTE_ADDR"), ServerVariable("REMOTE_ADDR"))) {
                    attackDetected(attack, score);
                    return ATTACK;
                }
            }
            else
                return ERROR;
            return OK;
        }

        public int checkURI()
        {
            var attack = "Vulnerability scanner in URL";
            var score = 10;

            if (IsSet(ServerVariable("REQUEST_URI")))
                using (var results = executeQuery("SELECT string FROM denyUrlString"))
                {
                    while (results.Read())
                        if (IncludesString(ServerVariable("REQUEST_URI"), (String)results["string"]))
                        {
                            attackDetected(attack, score);
                            return ATTACK;
                        }
                }
            else
                return ERROR;
            return OK;
        }

        public int checkHTTPVersion()
        {
            var attack = "Incorrect HTTP Version";
            var score = 100;
            var result = ServerVariable("SERVER_PROTOCOL");

            if (result != "") {
                if (result != "HTTP/1.1") {
                    attackDetected(attack, score);
                    return ATTACK;
                }
            }
            else
                return ERROR;
            return OK;
        }

        public int checkHostname(String hostname)
        {
            var attack = "Incorrect hostname";
            var score = 100;

            if (IsSet(hostname))
            {
                if (!IsSet(ServerVariable("HTTP_HOST")) || (ServerVariable("HTTP_HOST") != hostname))
                {
                    attackDetected(attack, score);
                    return ATTACK;
                }
            }
            else
                return ERROR;
            return OK;
        }

        public int checkNonExistingFile()
        {
            var attack = "Non existing file";
            var score = 5;
            var value = ServerVariable("REQUEST_URI");

            if (IsSet(value))
            {
                var path = System.Web.Hosting.HostingEnvironment.MapPath(value);
                var file = Path.GetFileName(path);
                var fileParts = file.Split('.');

                var extension = "";
                if (fileParts.Length > 1) 
                    extension = fileParts.Last<string>();

			    var results = executeQuery("SELECT extension FROM denyExtension");
                while (results.Read()) {
                    if (String.Equals(extension, (String) results["extension"])) {
                        attack = "Non existing backup file";
                        score = 100;
                        break;
                    }
                }

                attackDetected(attack, score);
                return ATTACK;
            }
            else
                return ERROR;
        }

        public int checkSpeed()
        {
            var time = GetTime();
            long value;

            if (!(SessionVariable("requests_last_minute") != null) || 
                ((long) SessionVariable("requests_last_minute") < (time.AddSeconds(-60).ToFileTimeUtc()))) {
                SessionVariableSet("requests_last_minute", time.ToFileTimeUtc());
                SessionVariableSet("amount_requests_last_minute", (long) 0);
            }

            // Increment request counter
            value = (long) SessionVariable("amount_requests_last_minute") + 1;
            SessionVariableSet("amount_requests_last_minute", value);
            // Maximum value reached
            if (value > MAX_REQUESTS_MINUTE) {
                attackDetected("Too many requests per minute", 100);
                return ATTACK;
            }

            return OK;
        }

        public int checkFakeCookie(string cookie_name = "admin", string cookie_value = "false")
        {
            var attack = "False cookie modified";
            var score = 100;

            if (IsSet(GetCookie(cookie_name)) && (GetCookie(cookie_name) != cookie_value)) {
                attackDetected(attack, score);
                return ATTACK;
            } 
		    else {
                SetCookie(cookie_name, cookie_value);
                return OK; // or error..
            }
        }

        public int checkFakeInput(String input, String value)
        {
		    var attack = "Fake input modified";
		    var score = 100;

            if (IsSet(input) && IsSet(value) && IsSet(RequestVariable(input)) && IsSet(SessionVariable(input)))
            {
                if (RequestVariable(input) != (String) SessionVariable(input)) {
                    attackDetected(attack, score);
                    return ATTACK;
                }
            }
            else
                return ERROR;
            return OK;
        }

        public int checkHTTPMethod(String method = "")
        {
		    var attack = "Incorrect HTTP method";
		    var score = 25;
            var value = ServerVariable("REQUEST_METHOD");

            if (IsSet(value) && method == "") {
		        var results = executeQuery("SELECT method FROM acceptHttpMethod");
		        var found = false;

                using (results)
                    while (results.Read())
                        if (String.Equals(value.ToLower(), ((String) results["method"]).ToLower())) { 
			                found = true;
                            break;
                        }

                if (found)
                    return OK;
		        else {
                    attack = "Blacklisted HTTP method";
                    attackDetected(attack, score);
                    return ATTACK;
                }
            }
            else if (IsSet(value) && (method != "")) {
                if (value != method) {
                    attackDetected(attack, score);
                    return ATTACK;
                }
            }
            else
		        return ERROR;
            return OK;
        }
    }
}

File: java/src/main/java/com/ioactive/defense/Defense.java
package com.ioactive.defense;

import static com.ioactive.defense.util.StringsListsUtils.containsIgnoreCase;
import static com.ioactive.defense.util.StringsListsUtils.isContainedIgnoringCase;

import java.io.File;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Enumeration;
import java.util.GregorianCalendar;
import java.util.List;
import java.util.Map;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;

import org.apache.commons.lang3.StringUtils;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import com.ioactive.defense.persistence.DefensePersistenceInterface;
import com.ioactive.defense.persistence.SQLiteDefensePersistence;
import com.ioactive.defense.util.ParameterMapSerializer;

public class Defense {

	public static final int OK = 1;
	public static final int ERROR = 0;
	public static final int ATTACK = -1;
	public static final int BAN = 100;
	public static final boolean DEBUG = true;
	public static final String DB = "attackers.db";
	public static final String NEWLINE = System.getProperty("line.separator");
	public static final String EMPTY = "";
	public static final String PROTOCOL = "HTTP/1.1";

	private static final DateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");

	private static final Map<String, String> SESSION_PARAMETERS = ImmutableMap.<String, String> of("user", "user", "REMOTE_ADDR", "ip", "cookie", "cookie");

	public DefensePersistenceInterface persistence = new SQLiteDefensePersistence();
	private Integer maximumRequestsPerMinute = 100;
	private Integer executionTimeWindow = 1; // Seconds
	private String cookieName = "admin";
	private String cookieValue = "false";
	private String fakeInputName = "_auth";
	private String fakeInputValue = "disabled";

	private String secretAdminAccount = "secrethiddenadminaccount";
	private String secretHiddenDirectory = "secrethiddendirectory";

	public Defense() {
	}

	public Defense(DefensePersistenceInterface defensePersistenceInterface) {
		persistence = defensePersistenceInterface;
	}

	public void setPersistence(DefensePersistenceInterface persistence) {
		this.persistence = persistence;
	}

	public void setMaximumRequestsPerMinute(Integer maximumRequestsPerMinute) {
		this.maximumRequestsPerMinute = maximumRequestsPerMinute;
	}

	public void setExecutionTimeWindow(Integer executionTimeWindow) {
		this.executionTimeWindow = executionTimeWindow;
	}

	public void setCookieName(String cookieName) {
		this.cookieName = cookieName;
	}

	public void setCookieValue(String cookieValue) {
		this.cookieValue = cookieValue;
	}

	public void setFakeInputName(String fakeInputName) {
		this.fakeInputName = fakeInputName;
	}

	public void setFakeInputValue(String fakeInputValue) {
		this.fakeInputValue = fakeInputValue;
	}

	public void setSecretAdminAccount(String secretAdminAccount) {
		this.secretAdminAccount = secretAdminAccount;
	}

	public void setSecretHiddenDirectory(String secretHiddenDirectory) {
		this.secretHiddenDirectory = secretHiddenDirectory;
	}

	/*
	 * metodos pendientes 1- function checkFakeCookie($cookie_name = "admin", $cookie_value = "false") 2 - function
	 * checkFakeInput($input, $value) 3 - function checkSpeed() 4 - function exception_handler($exception) 5-
	 * logoutSession (esta solo definido) 6 - function isAttacker() 7 - function nonExistingFile()
	 */

	/**
	 * Check that the user is using the correct HTTP method
	 * 
	 * @param request
	 * @param checkMethod
	 * @return
	 */
	public int checkHttpMethod(HttpServletRequest request, String checkMethod) {
		String attack = "Incorrect HTTP method";
		Integer score = 25;

		String method = request.getMethod();

		if (StringUtils.isBlank(method)) {
			return ERROR;
		}

		if (StringUtils.isBlank(checkMethod)) {
			List<String> results = persistence.getAcceptedMethods();
			if (containsIgnoreCase(results, method)) {
				return OK;
			} else {
				attack = "Blacklisted HTTP method";
				attackDetected(request, attack, score);
				return ATTACK;
			}
		} else {
			if (checkMethod.equalsIgnoreCase(method)) {
				return OK;
			} else {
				attackDetected(request, attack, score);
				return ATTACK;
			}
		}
	}

	/**
	 * Check if the URL contains a string flagged as an attacker
	 * 
	 * @param request
	 * @return
	 */
	public int checkURI(HttpServletRequest request) {
		String attack = "Vulnerability scanner in URL";
		int score = 10;

		String requestURI = request.getRequestURI();

		if (StringUtils.isBlank(requestURI))
			return ERROR;

		List<String> deniedURLs = persistence.getDenyURLs();
		if (isContainedIgnoringCase(deniedURLs, requestURI)) {
			attackDetected(request, attack, score);
			return ATTACK;
		}
		return OK;
	}

	/**
	 * 
	 * @param request
	 * @return
	 */
	public int checkUserAgent(HttpServletRequest request) {
		String attack = "Vulnerability scanner in user-agent";
		String requestUserAgent = request.getHeader("User-Agent");
		String sessionUserAgent = (String) request.getSession().getAttribute("HTTP_USER_AGENT");
		int score = 100;

		if (requestUserAgent == null) {
			return ERROR;
		}

		List<String> deniedUserAgents = persistence.getDeniedUserAgents();
		if (containsIgnoreCase(deniedUserAgents, requestUserAgent)) {
			attackDetected(request, attack, score);
			return ATTACK;
		}

		if (sessionUserAgent != null && requestUserAgent != null && !sessionUserAgent.equals(requestUserAgent)) {
			attack = "User-agent changed during user session";
			attackDetected(request, attack, score);
			return ATTACK;
		}
		return OK;
	}

	public int checkHTTPVersion(HttpServletRequest request) {
		String attack = "Incorrect HTTP Version";
		String protocol = request.getProtocol();
		int score = 100;

		if (protocol != null) {
			if (!protocol.equals(PROTOCOL)) {
				attackDetected(request, attack, score);
				return ATTACK;
			}
		} else
			return ERROR;
		return OK;
	}

	public int checkConcurrentSession(HttpServletRequest request) {
		String attack = "The IP address of the user changed for the cookie";
		int score = 25;

		String remoteAddress = request.getRemoteAddr();
		String sessionRemoteAddress = (String) request.getSession().getAttribute("REMOTE_ADDR");

		if (remoteAddress != null && sessionRemoteAddress != null) {
			if (!remoteAddress.equals(sessionRemoteAddress)) {
				attackDetected(request, attack, score);
				return ATTACK;
			}
		} else
			return ERROR;
		return OK;
	}

	public int checkHostname(HttpServletRequest request) {
		String attack = "Incorrect hostname";
		int score = 100;

		String requestHostName = request.getServerName();
		String sessionHostName = (String) request.getSession().getAttribute("HTTP_HOST");

		if (requestHostName != null) {
			if (sessionHostName == null) {
				sessionHostName = requestHostName;
				request.getSession().setAttribute("HTTP_HOST", requestHostName);
			}
			if (requestHostName == null || !requestHostName.equals(sessionHostName)) {
				attackDetected(request, attack, score);
				return ATTACK;
			}
		} else
			return ERROR;

		return OK;
	}

	/**
	 * Define a cookie to the user, that in case it is changed, it will be flagged as an attacker
	 * 
	 * @return
	 */
	public int checkFakeCookie(HttpServletRequest request, HttpServletResponse response) {
		String attack = "False cookie modified";
		int score = 100;

		Cookie[] cookies = request.getCookies();
		Cookie cookie = null;
		for (Cookie cookieItem : cookies) {
			if (StringUtils.equals(cookieItem.getName(), cookieName)) {
				cookie = cookieItem;
				break;
			}
		}

		if (cookie != null && !StringUtils.equals(cookie.getValue(), cookieValue)) {
			attackDetected(request, attack, score);
			return ATTACK;
		} else {
			response.addCookie(new Cookie(this.cookieName, this.cookieValue));
			return OK; // or error..
		}

	}

	/**
	 * Define a fake input field in a form and save the fake value in the session. If it is changed, consider it an
	 * attack.
	 * 
	 * @return
	 */
	public int checkFakeInput(HttpServletRequest request, String input, String value) {
		String attack = "Fake input modified";
		int score = 100;

		if (StringUtils.isEmpty(input))
			return ERROR;
		String inputValue = request.getParameter(input);

		if (!StringUtils.equals(value, inputValue)) {
			attackDetected(request, attack, score);
			return ATTACK;
		}
		return OK;
	}

	/**
	 * Define a fake input field in a form and save the fake value in the session. If it is changed, consider it an
	 * attack. This one uses the lib's set input names and values
	 * 
	 * @return
	 */
	public int checkFakeInput(HttpServletRequest request) {
		String attack = "Fake input modified";
		int score = 100;

		String inputValue = request.getParameter(fakeInputName);
		if (!StringUtils.equals(fakeInputValue, inputValue)) {
			attackDetected(request, attack, score);
			return ATTACK;
		}
		return OK;
	}

	/**
	 * Check how many requests per minute the user sends
	 * 
	 * @return
	 */
	public int checkSpeed(HttpServletRequest request) {
		String attack = "Too many requests";
		int score = 100;

		Date now = new Date();
		Calendar calendar = new GregorianCalendar();
		calendar.setTime(now);
		calendar.add(Calendar.MINUTE, -1);

		HttpSession session = request.getSession();
		Date requestLastMinute = (Date) session.getAttribute("requests_last_minute");

		if (requestLastMinute == null || requestLastMinute.before(calendar.getTime())) {
			session.setAttribute("requests_last_minute", now);
			session.setAttribute("amount_requests_last_minute", Integer.valueOf(0));
		}

		Integer requestCountForTheLastMinute = (Integer) session.getAttribute("amount_requests_last_minute");
		requestCountForTheLastMinute++;
		if (requestCountForTheLastMinute > maximumRequestsPerMinute) {
			attackDetected(request, attack, score);
			return ATTACK;
		}
		session.setAttribute("amount_requests_last_minute", requestCountForTheLastMinute);
		return OK;
	}

	/**
	 * Catch unhandled exceptions
	 * 
	 * @return
	 */
	public int exceptionHandler(HttpServletRequest request, Exception exception) {
		String attack = "Uncaught exception.";
		Integer score = 100;

		String attackDescription = String.format("%s Exception message: %s", attack, exception.getMessage());
		attackDetected(request, attackDescription, score);
		return ATTACK;
	}

	/**
	 * If tries to get to inexistante file
	 * 
	 * @return
	 */
	public int nonExistingFile() {
		// no sense for java
		return OK;
	}

	public int checkPath(HttpServletRequest request, String file) {
		File f = new File(file);
		try {
			String absoluteFile = f.getAbsolutePath();
			String canonicalFile = f.getCanonicalPath();
			if (!absoluteFile.equals(canonicalFile)) {
				attackDetected(request, "Path traversal detected", 100);
				return ATTACK;
			} else
				return OK;
		} catch (Exception e) {
			return ERROR;
		}
	}

	public int checkFakeSecretAdminAccountLeakage(HttpServletRequest request, String responseBody) {
		if (responseBody == null) {
			return ERROR;
		}
		if (responseBody.contains(secretAdminAccount)) {
			attackDetected(request, "Passwords leaked", 100);
			return ATTACK;
		}
		return OK;

	}

	public int checkFakeSecretHiddenDirectoryLeakage(HttpServletRequest request, String responseBody) {
		if (responseBody == null) {
			return ERROR;
		}
		if (responseBody.contains(secretHiddenDirectory)) {
			attackDetected(request, "Files leaked", 100);
			return ATTACK;
		}
		return OK;
	}

	public int checkExecutionTime(HttpServletRequest httpRequest, Date startTime) {
		if (startTime == null) {
			return ERROR;
		}
		Calendar c = new GregorianCalendar();
		c.add(Calendar.SECOND, executionTimeWindow * -1);
		if ((startTime.before(c.getTime()))) {// if it takes more than 1 seconds..
			attackDetected(httpRequest, "Too much time", 20);
			return ATTACK;
		}
		return OK;
	}

	// ---------------------------- common methods ----------------------------
	/**
	 * Save the attack...
	 * 
	 * @param request
	 * @param attack
	 * @param score
	 */
	public void attackDetected(HttpServletRequest request, String attack, int score) {

		Map<String, Object> sessionParameters = getSessionParameters(request.getSession());
		String serializedParameterMap = ParameterMapSerializer.serializeParameterMap(request.getParameterMap());
		// String serializedParameterMap = Dumper.dump(request.getParameterMap());

		Map<String, Object> args = Maps.newLinkedHashMapWithExpectedSize(10);
		args.put("timestamp", DATE_FORMAT.format(new Date())); // better than: args.put("timestamp", new Date());
		args.put("application", "test-defense");
		args.put("ip", sessionParameters.get("ip"));
		args.put("user", sessionParameters.get("user"));
		args.put("cookie", sessionParameters.get("cookie"));
		args.put("uri", request.getRequestURI());
		args.put("parameter", serializedParameterMap);
		args.put("attack", attack);
		args.put("score", score);

		persistence.logAttack(args);
		logoutSession(request);

		StringBuilder alertInfoBuilder = new StringBuilder();
		alertInfoBuilder.append("The last attack from the user was: ").append(attack);
		if (score >= BAN) {
			alertInfoBuilder.append(". The user was automatically mark as an attacker.");
		} else {
			alertInfoBuilder.append(". The user was mark as an attacker because of a series of events.");
		}
		alertInfoBuilder.append(NEWLINE);

		alertInfoBuilder.append("Attacker details:").append(NEWLINE);
		alertInfoBuilder.append("IP: ").append(args.get("ip")).append(NEWLINE);
		alertInfoBuilder.append("User: ").append(args.get("user")).append(NEWLINE);
		alertInfoBuilder.append("Cookie: ").append(args.get("cookie")).append(NEWLINE);
		// No sense to have this in java... perhaps the stack might help little more.
		// alertInfoBuilder.append("File: N/A").append(NEWLINE);
		alertInfoBuilder.append("URI: ").append(args.get("uri")).append(NEWLINE);
		alertInfoBuilder.append("Parameter: ").append(args.get("parameter")).append(NEWLINE);

		alertAdmin(alertInfoBuilder.toString());
	}

	/**
	 * Logout user and destroy the session
	 */
	private void logoutSession(HttpServletRequest request) {
		// TODO Auto-generated method stub
	}

	/**
	 * Get the session stuff: IP, user (optional), cookie (optional)
	 * 
	 * @param session
	 * @return
	 */
	private Map<String, Object> getSessionParameters(HttpSession session) {
		Map<String, Object> sessionParameters = Maps.newLinkedHashMap();
		Enumeration<String> attributeNames = session.getAttributeNames();
		while (attributeNames.hasMoreElements()) {
			String attributeName = attributeNames.nextElement();
			if (SESSION_PARAMETERS.containsKey(attributeName)) {
				sessionParameters.put(SESSION_PARAMETERS.get(attributeName), session.getAttribute(attributeName));
			}
		}
		return sessionParameters;
	}

	/**
	 * Provide an alert
	 * 
	 * @param alertInfo
	 */
	public void alertAdmin(String alertInfo) {
		if (DEBUG)
			System.out.println(String.format("%s%s", alertInfo, NEWLINE));
	}

}